### PR TITLE
[WIP] Add pasqal device

### PR DESCRIPTION
**Pull request by co9olguy**
Adds a new device "cirq.pasqal", which uses the `cirq.pasqal.PasqalDevice` simulator as the execution backend. Cirq should be installed from the following fork: https://github.com/lhenriet/Cirq

Notes
-----

- Pasqal devices require `ThreeDGridQubit`s, which are qubits which have positions in three dimensions.
- Based on reading the plugin code, there are a number of places where an implicit assumption is made that the qubits are `LineQubit`:
    - in initialization of cirq_device.py
    - in _apply_basis_state in simulator_device.py
    - in _apply_qubit_state_vector in simulator_device.py
    - in apply in simulator_device.py
- It should be relatively straightforward to allow other qubit types (e.g., `GridQubit` or the `ThreeDGridQubit`s needed for the `PasqalDevice`, but it would require some small updates to the plugin to support more general cases than `LineQubit`. In particular, I identified that wherever `len` is called on the qubits, it will fail unless the qubits are `LineQubit`s.
- Fortunately, Cirq provides total ordering methods, so we can unambiguously assign every qubit in a grid a reproducible integer number, which allows us to map from PL's linear wire naming to the other schemes
- `PasqalDevice` also requires an initialization argument `control_radius`. Two qubits in the grid can undergo two-qubit gates iff they are within `control_radius` of one another

With the addition of better handling for different qubit arrangements, I think we can relatively easily add the `PasqalDevice`.

One option for the future is to perhaps expose the positions of the qubits in 3D space as a variable that Pennylane can tune and optimize, but I think it might be tricky at the moment, since it would require updating the qubits after the device is initialized (and I'm not exactly sure whether Cirq/Pasqal supports non-static arrangements yet)

### CODE REVIEW

**Comment by josh146** [comment]
@co9olguy, I have a couple of questions you might be able to answer on this PR:

> but it would require some small updates to the plugin to support more general cases than `LineQubit`. In particular, I identified that wherever len is called on the qubits, it will fail unless the qubits are `LineQubits`.

From what I can tell, `qubits` accepts a list of `ThreeDGridQubits`, so length will always be well defined? e.g.,

```python
qubits = [
	pasqal.ThreeDGridQubit(0, 0, 0),
	pasqal.ThreeDGridQubit(0, 1, 0),
	pasqal.ThreeDGridQubit(1, 0, 0),
	pasqal.ThreeDGridQubit(1, 1, 0)
]

dev = qml.device("cirq.pasqal", wires=4, shots=1000, analytic=True, qubits=qubits, control_radius=0.1)

@qml.qnode(dev, interface="tf")
def circuit(params):
	qml.templates.BasicEntanglerLayers(params, wires=[0, 1, 2, 3])
	return qml.expval(qml.PauliZ(0) @ qml.PauliX(2))
```

works fine for me without the `isinstance(qubit, LineQubit) and ...` validation.

> Fortunately, Cirq provides total ordering methods, so we can unambiguously assign every qubit in a grid a reproducible integer number, which allows us to map from PL's linear wire naming to the other schemes

This one follows from above; it seems that the `len(qubits)` works for sequences of `ThreeDGridQubits`, so an internal mapping shouldn't be needed?

> One option for the future is to perhaps expose the positions of the qubits in 3D space as a variable that Pennylane can tune and optimize, but I think it might be tricky at the moment, since it would require updating the qubits after the device is initialized

This might actually be doable in a way; e.g.,

```python
with tf.GradientTape() as tape:
    res = circuit(params)
    dev.qubits = pasqal.ThreeDGridQubit(*func(res))
```

The only hurdle I can see though is that `ThreeDGridQubit` accepts only _integers_ for the coordinates, not floats, so will not be trainable for now.

**Comment by co9olguy** [comment]
> The main hurdle I can see though is that `ThreeDGridQubit` accepts only _integers_ for the coordinates, not floats

Where did you see this?

**Comment by co9olguy** [comment]
> From what I can tell, `qubits` accepts a list of `ThreeDGridQubits`, so length will always be well defined? e.g.,
> 
> ```python
> qubits = [
>     pasqal.ThreeDGridQubit(0, 0, 0),
>     pasqal.ThreeDGridQubit(0, 1, 0),
>     pasqal.ThreeDGridQubit(1, 0, 0),
>     pasqal.ThreeDGridQubit(1, 1, 0)
> ]
> 
> dev = qml.device("cirq.pasqal", wires=4, shots=1000, analytic=True, qubits=qubits, control_radius=0.1)
> 
> @qml.qnode(dev, interface="tf")
> def circuit(params):
>     qml.templates.BasicEntanglerLayers(params, wires=[0, 1, 2, 3])
>     return qml.expval(qml.PauliZ(0) @ qml.PauliX(2))
> ```
> 
> works fine for me without the `isinstance(qubit, LineQubit) and ...` validation.

I think what happened (forgive me, been a couple weeks) is I eventually figured it out (you need to pass a list of qubits rather than a single qubit object), but I didn't cross that out of my working notes above before posting here

**Comment by co9olguy** [comment]
> > From what I can tell, `qubits` accepts a list of `ThreeDGridQubits`, so length will always be well defined? e.g.,
> > ```python
> > qubits = [
> >     pasqal.ThreeDGridQubit(0, 0, 0),
> >     pasqal.ThreeDGridQubit(0, 1, 0),
> >     pasqal.ThreeDGridQubit(1, 0, 0),
> >     pasqal.ThreeDGridQubit(1, 1, 0)
> > ]
> > 
> > dev = qml.device("cirq.pasqal", wires=4, shots=1000, analytic=True, qubits=qubits, control_radius=0.1)
> > 
> > @qml.qnode(dev, interface="tf")
> > def circuit(params):
> >     qml.templates.BasicEntanglerLayers(params, wires=[0, 1, 2, 3])
> >     return qml.expval(qml.PauliZ(0) @ qml.PauliX(2))
> > ```
> > 
> > 
> > works fine for me without the `isinstance(qubit, LineQubit) and ...` validation.
> 
> I think what happened (forgive me, been a couple weeks) is I eventually figured it out (you need to pass a list of qubits rather than a single qubit object), but I didn't cross that out of my working notes above before posting here

**Comment by josh146** [comment]
> Where did you see this?

The `ThreeDGridQubit` accepts only integers for the coordinates, as indicated by type hints and docstrings:

https://github.com/lhenriet/Cirq/blob/125722761ca56a641db7e15c89303ed55f076441/cirq/pasqal/pasqal_qubits.py#L20

There's no specific validation _stopping_ us from initializing the grid with floats instead, but it might not work end-to-end, since it looks like integers are assumed? It gets confusing though, since the `PasqalDevice` class sometimes assumes that these values are floats.

> @josh146 did you actually evaluate the QNode? 😀

Almost! Everything seemed to work all the way down to `device.execute()` was called. It fails with `PasqalDevice` has no `simulate` method. I couldn't quite work out how the `Pasqaldevice` was to be executed

**Comment by co9olguy** [comment]
> The `ThreeDGridQubit` accepts only integers for the coordinates, as indicated by type hints and docstrings:
> 
> https://github.com/lhenriet/Cirq/blob/125722761ca56a641db7e15c89303ed55f076441/cirq/pasqal/pasqal_qubits.py#L20

Interesting, this looks like a recent change (less than 3 weeks ago). Wonder why? :thinking: 

> Almost! Everything seemed to work (applying etc.) until `device.execute()` was called. It fails with `PasqalDevice` has no `simulate` method. I couldn't quite work out how the `Pasqaldevice` was to be executed

Got everything to work now. I just needed to take some time to understand better how Cirq implements things (e.g., `Simulator` and `Device` classes have completely different roles)

**Comment by co9olguy** [comment]
@josh146 looks like some of the tests are failing because of that cirq versioning bug :frowning:

File: requirements.txt [hunk 1]
```diff
@@ -1,3 +1,4 @@
 pennylane>=0.9
-cirq>=0.7
+google-api-core[grpc]<=1.14.0
+git+https://github.com/lhenriet/Cirq.git
```
[the remark this answers is not in the document]
Comment (co9olguy) [reply requirements.txt:3]: Let's keep this in a separate branch, and undocumented, until that fork is successfully merged in. I'll try to keep abreast of when that might happen. In the meantime, we can continue our next steps (developing a tutorial) using the holding branch

File: pennylane_cirq/pasqal_device.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    short_name = "cirq.pasqal"
+
+    def __init__(self, wires, shots=1000, analytic=True, qubits=None, control_radius=1.0):
+
+        if not qubits:
+            qubits = [pasqal.ThreeDGridQubit(wire, 0, 0) for wire in range(wires)]
+        self.control_radius = float(control_radius)
+        if self.control_radius < 0:
+            raise ValueError("The control_radius must be a non-negative real number.")
+        super().__init__(wires, shots, analytic, qubits)
```
[the remark this answers is not in the document]
Comment (co9olguy) [reply pennylane_cirq/pasqal_device.py:52]: Yes it did!

**Pull request merged**